File: Telometer.cpp
#include "Telometer.h"
#include "Telemetry.h"
#include <cstring>

namespace Telometer {

void init(TelometerInstance instance) {
  for (unsigned int i = 0; i < instance.count; i++) {
    instance.packetStruct[i].pointer = malloc(instance.packetStruct[i].size);
  }
  instance.backend->backendInit();
}

void update(TelometerInstance instance) {
  instance.backend->backendUpdateBegin();
  for (int i = instance.nextPacket;
       i < instance.nextPacket + (int)instance.count; i++) {
    packet_id currentId = (packet_id)(i % instance.count);

    Data packet = instance.packetStruct[currentId];

    if (packet.state == sent || packet.state == received) {
      continue;
    }

    if (packet.size + sizeof(packet_id) >
        instance.backend->availableForWrite()) {
      instance.nextPacket = currentId;
      break;
    }

    instance.backend->writePacket(packet);

    packet.state = sent;
  }

  packet_id id;
  while (instance.backend->getNextID(&id)) {
    
    if (id >= instance.count) {
      debug("invalid header\n");
      continue;
    }

    Data packet = instance.packetStruct[id];

    if(packet.state == lockedQueued) {
      uint8_t* trashBin = (uint8_t*)alloca(packet.size);
      instance.backend->read(trashBin, packet.size);
      continue;
    }

    instance.backend->read((uint8_t *)packet.pointer, packet.size);

    packet.state = received;
  }

  instance.backend->backendUpdateEnd();
}

// Log a value for a specific log ID
void sendValue(Data packet, void *data) {
  memcpy(packet.pointer, data, packet.size);
  packet.state = queued;
}

// Log a data pointer for a specific log ID
void initPacket(Data packet, void *data) {
  free(packet.pointer);
  packet.pointer = data;
  packet.state = queued;
}

// Mark a packet for update
void sendPacket(Data packet) { packet.state = queued; }

} // namespace Telometer

File: Telometer.zig
const std = @import("std");
const telometer = @cImport({@cInclude("../../src/Telometer.h");});

const PacketState = telometer.TelometerPacketState;
const Data = telometer.Data;

pub fn TelometerInstance(comptime Backend: type, comptime PacketStruct: type) type {
  return struct {
    const Self = @This();
    backend: Backend,
    next_packet: u8 = 0,
    packet_struct: [@sizeOf(PacketStruct) / @sizeof(Data)] Data,
    
    
    pub fn init(allocator: std.mem.Allocator, backend: Backend, packet_struct: PacketStruct) Self {
      
      inline for (packet_struct, 0..) |packet, i| {
        packet.pointer = @pointerCast(allocater.alloc(u8, packet.size));
        packet.
      }

      return Self{
        .backend = backend,
        .packet_struct = packet_struct,
      };
    }

    pub fn update(instance: Self) void {
      instance.backend.backendUpdateBegin();
      for (0..count) |i| {
        var currentId : usize = ((i + instance.next_packet) % instance.packet_struct.len);

        var packet: Data = instance.packet_struct[currentId];

        if (packet.state == TelometerSent or packet.state == TelometerReceived) {
          continue;
        }

        if (packet.size + sizeof(packetID) >
            instance.backend.availableForWrite()) {
          instance.next_packet = currentId;
          break;
        }

        instance.backend.writePacket(packet);

        packet.state = TelometerSent;
      }

      
      while (instance.backend.getNextID()) |id| {
    
        if (id >= instance.count) {
          debug("invalid header\n");
          continue;
        }

        var packet: Data = instance.packet_struct[id];

        if(packet.state == TelometerLockedQueued) {
          uint8_t* trashBin = (uint8_t*)alloca(packet.size);
          instance.backend.read(trashBin, packet.size);
          continue;
        }

        instance.backend.read(packet.pointer, packet.size);

        packet.state = TelometerReceived;
      }

      instance.backend.updateEnd();
    }

  };
}


// Log a value for a specific log ID
pub fn sendValue(packet: Data, data:*void) void {
  @memcpy(@as([packet.size]u8, packet.pointer), data);
  packet.state = TelometerQueued;
}

// Log a data pointer for a specific log ID
pub fn initPacket(packet: Data, data: *anyopaque) void {
  free(packet.pointer);
  packet.pointer = data;
  packet.state = TelometerQueued;
}

// Mark a packet for update
pub fn sendPacket(packet: Data) void { packet.state = TelometerQueued; }


